File: Code/src/portfoliomanager/model/Account.java
package portfoliomanager.model;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;

/**
 * The Account Class.
 * @author Sam
 * @version Spring 2025
 */
public class Account {
	private String username;
	private String password;
	private String auth;
	private List<Holding> holdings;
	private double fundsAvailable;
	
	/**
	 * Instantiates a new account.
	 *
	 * @param username the username
	 * @param password the password
	 * @param auth the token for the account
	 */
	public Account(String username, String password, String auth) {
		if (username.isBlank() || password.isBlank() || auth.isBlank()) {
			throw new IllegalArgumentException("Username or password don't meet specified requirements.");
		}
		this.auth = auth;
		this.username = username;
		this.password = password;
		this.holdings = new ArrayList<Holding>();
	}
	
	/**
	 * Instantiates a new account
	 * @param username the username
	 * @param password the password
	 * @param auth the auth token
	 * @param fundsAvailable the funds available to the account
	 * @post this.fundsAvaiable == fundsAvailable
	 */
	public Account(String username, String password, String auth, double fundsAvailable) {
		this(username, password, auth);
		this.fundsAvailable = fundsAvailable;
	}
	
	/**
	 * Gets the user name.
	 *
	 * @return the user name
	 */
	public String getUserName() {
		return this.username;
	}

	/**
	 * Sets the user name.
	 *
	 * @param newName the new user name
	 */
	public void setUserName(String newName) {
		if (newName.isBlank()) {
			throw new IllegalArgumentException("Username doesn't meet specified requirements.");
		}
		
		this.username = newName;
	}
	
	/**
	 * Gets the password.
	 *
	 * @return the password
	 */
	public String getPassword() {
		return this.password;
	}
	
	/**
	 * Sets the password given the old, current password and the new password to be changed to.
	 *
	 * @param oldP the old P
	 * @param newP the new P
	 */
	public void setPassword(String oldP, String newP) {
		if (oldP != this.password) {
			throw new IllegalArgumentException("The old password is incorrect.");
		}
		
		if (newP.isBlank()) {
			throw new IllegalArgumentException("The new password doesn't meet requirements.");
		}
		
		this.password = newP;
	}
	
	/**
	 * Gets the holding of the user
	 * @return the holding
	 */
	public List<Holding> getHoldings() {
		return this.holdings;
	}
	
	/**
	 * Adds the holding of the user 
	 * @param holding the holding to add
	 * @return true or false based on if added properly
	 */
	public boolean addHolding(Holding holding) {
		for (Holding currHolding : this.holdings) {
			if (currHolding != null && currHolding.getName().equals(holding.getName())) {
				currHolding.setAmountHeld(holding.getAmountHeld());
				return true;
			}
		}
		this.holdings.add(holding);
		
		return true;
	}
	
	/**
	 * Gets the funds available 
	 * @return the funds
	 */
	public double getFundsAvailable() {
		return this.fundsAvailable;
	}
	
	/**
	 * Sets the funds available
	 * @param amount the amount to set
	 */
	public void setFundsAvailable(double amount) {
		this.fundsAvailable = this.roundToTwoDecimalPlaces(amount);
	}
	
	private double roundToTwoDecimalPlaces(double value) {
		BigDecimal bd = new BigDecimal(value);
		bd = bd.setScale(2, RoundingMode.HALF_UP);
		return bd.doubleValue();
	}
	
	/**
	 * Gets the accounts auth token
	 * @return the auth token
	 */
	public String getAuth() {
		return this.auth;
	}
	
	/**
	 * Sets the accounts auth token.
	 *
	 * @param authtoken the new authtoken
	 */
	public void setAuth(String authtoken) {
		this.auth = authtoken;
	}
	/**
	 * Sets the users holdings
	 * @param holdings the holdings to set
	 * @post this.holdings == holdings
	 */
	
	public void setHoldings(List<Holding> holdings) {
		this.holdings = holdings;
	}
	
	@Override
	public boolean equals(Object object) {
		Account other = (Account) object;
		boolean equal = false;
		if (this.username.equals(other.getUserName())) {
			if (this.auth.equalsIgnoreCase(other.getAuth())) {
				equal = true;
			}
		}
		return equal;
	}

}

File: Code/src/portfoliomanager/model/Account.py
'''
Created on Apr 3, 2025

@author: Colby
'''
from model import Holding
import bcrypt
'''
The account class
'''
class Account:
    '''
    Instanties an instance of an account
    @pre username and password is not empty
    @post self.username == username, self.password_hash == random hashed value
          self.holdings is empty, self.funds_available == 0
    '''
    def __init__(self, username, password):
        if not username.strip() or not password.strip():
            raise ValueError("Username or password don't meet specified requirements.")

        self.username = username
        self.password_hash = self.hash_password(password)
        self.holdings = []
        self.funds_available = 0.0
    
    def hash_password(self, password: str) -> bytes:
        """Hashes a password using bcrypt."""
        return bcrypt.hashpw(password.encode(), bcrypt.gensalt())

    def verify_password(self, password: str) -> bool:
        """Verifies the provided password against the stored hash."""
        return bcrypt.checkpw(password.encode(), self.password_hash)

    def modify_holding(self, holding: Holding, add=True) -> bool:
        '''
        Adds a holding to the account
        @post holding is in holdings
        '''
        for curr_holding in self.holdings:
            if curr_holding.name == holding.name:
                if(add):
                    curr_holding.amount += float(holding.amount)
                else:
                    curr_holding.amount -= float(holding.amount)
                
                holding.set_amount_held(curr_holding.amount)
                return True
                

        self.holdings.append(holding)
        return True
    
    def modify_funds(self, totalCost: float, add= True):
        if (add):
            self.funds_available += float(totalCost)
        else:
            self.funds_available -= float(totalCost)
        
    def get_holding(self, name):
        '''
        Gets the holding in the account given the name
        @returns the holding with that name
        '''
        for holding in self.holdings:
            if (holding.name == name):
                return holding
    def change_password(self, current_password: str, new_password: str):
        """Changes the account password after verifying the current one."""
        if not new_password.strip():
            raise ValueError("New password must not be empty.")
        
        if not self.verify_password(current_password):
            raise ValueError("Current password is incorrect.")
        
        self.password_hash = self.hash_password(new_password)
    def __eq__(self, other):
        '''
        Overrides the == operator with the following'''
        if isinstance(other, Account):
            return self.username == other.username